File: 0201-0300/0207-Course Schedule/0207-Course Schedule.cpp
// BFS
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> degrees(numCourses);
        for (auto& p : prerequisites) {
            graph[p[1]].push_back(p[0]);
            ++degrees[p[0]];
        }

        queue<int> Q;        
        for (int i = 0; i < numCourses; ++i) {
            if (degrees[i] == 0) {
                Q.push(i);
            }
        }
        
        int count = 0;
        while (!Q.empty()) {
            int i = Q.front();
            ++count;
            Q.pop();
            for (int j : graph[i]) {
                if (--degrees[j] == 0) {
                    Q.push(j);
                }
            }
        }
        
        return count == numCourses;
    }
};

// DFS
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        for (auto& p : prerequisites) {
            graph[p[1]].push_back(p[0]);
        }

        vector<int> visited(numCourses);
        for (int i = 0; i < numCourses; ++i) {
            if (!dfs(graph, i, visited)) {
                return false;
            }
        }
        
        return true;
    }
    
private:
    bool dfs(vector<vector<int>>& graph, int course, vector<int>& visited) {
        if (visited[course] == 1) {
            return true;
        }
        else if (visited[course] == 2) {
            return false;
        }
        
        visited[course] = 2;
        for (int i : graph[course]) {
            if (!dfs(graph, i, visited)) {
                return false;
            }
        }
        
        visited[course] = 1;
        return true;
    }
};

File: 0201-0300/0207-Course Schedule/0207-Course Schedule.py
import collections
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = collections.defaultdict(list)
        degrees = [0] * numCourses
        for c1, c2 in prerequisites:
            graph[c2].append(c1)
            degrees[c1] += 1
                
        Q = collections.deque([c for c in range(numCourses) if degrees[c] == 0])
        count = 0
        while Q:
            count += 1
            c = Q.popleft()
            for n in graph[c]:
                degrees[n] -= 1
                if degrees[n] == 0:
                    Q.append(n)
                    
        return count == numCourses
